design.md

```diff
@@ -0,0 +1,109 @@
+
+## Motivating use case ##
+
+Interest in data science is exploding.  As a result, many individuals across disciplines and with diverse computational backgrounds are discovering data science tools.  At schools like UC Berkeley (Cal), these individuals include students and instructors at all levels:  from incoming freshman to students across the college, masters, PhD, and MBA students, postdocs, lecturers, junior and senior faculty.  Instructors across the campus are developing data-centric classes, from freshman seminars to advanced special topics courses in areas such as history, literature, ecology, neuroscience, and health.  For example, here is a list of 11 data science seminars, aimed at freshmen, being taught this spring (2016) at Cal: https://data-8.appspot.com/sp16/modules/extra_tabs/render?index=3
+Many of these instructors do not have traditional training in computer science or statistics, and are learning data science skills on the fly, as they develop and teach their courses.
+
+Jupyter notebooks are widely employed in classes at Cal because they provide a natural environment for introducing data science skills to students.  Critically, they enable a browser-based interface to computation in the cloud, meaning that students only need a browser to start programming and interacting with data.  
+
+However, the tools currently available for instructors require considerably more experience with computational and programming tools. In order to include and support this multidisciplinary group of instructors who are developing a rich ensemble of classes, particularly from disciplines currently underrepresented in data science, the tools available must be designed to facilitate the work of instructors whose research and educational backgrounds do not include, for instance, tools for version control. We focus here on a particular problem that they encounter early in developing and teaching the courses: managing the notebooks, data, and other files they develop and distribute to students.
+
+For the instructors' current workflow, see https://github.com/data-8/connector-instructors#workflow 
+
+# Summary of Solution #
+
+Implement Dropbox-like synchronization of specific folders, accessible via the Jupyter Notebook interface. We will initially provide a GitHub backed implementation, but in the future other backends could be made available (e.g. Dropbox, generic git). We will explicitly avoid making a true Git GUI client - the user's mental model should be 'this is similar to Dropbox.'
+
+# Implementation details #
+
+This would be implemented as a Jupyter notebook backend extension. We'd use a very small subset of git commands, tied specifically into GitHub, to provide this synchronizing functionality.
+
+## Synchronization ##
+
+The actual synchronization would be handled by the python part of the nbextension, via logic similar to https://gist.github.com/yuvipanda/6cdeab86ccaa15687b25. 
+
+1. Every 30s (or other configurable interval)...
```

Comment:
Actually, this may be a problem for `git pull` when people receive files since Github only allows 60 requests per hour for unauthenticated requests. In addition, I think that in this case the requests from all the people receiving the files would count towards that limit.

Eg. Instructor adds file to sync'ed folder. Persons A, B, C `git pull` the repo. Now we have 57 requests left because all the requests came from the same IP. If we pull once every 30 seconds, we'll run out of requests in ~10 minutes.

If persons A, B, C use the instructor's token, then we'll still run out of requests if > 40 people get synced every 30 seconds.

If each person uses their own Github token we'd be fine, but that requires students to make a Github account and such. This doesn't seem like something we want to do.